alternator: avoid oversized allocation in Query/Scan

nyh opened the pull request:
This series fixes one cause of oversized allocations - and therefore potentially stalls and increased tail latencies - in Alternator.

The first patch in the series is the main fix - the later patches are cleanups requested by reviewers but also involved other pre-existing code, so I did those cleanups as separate patches.

Alternator's Scan or Query operation return a page of results. When the number of items is not limited by a "Limit" parameter, the default is to return a 1 MB page. If items are short, a large number of them can fit in that 1MB. The test test_query.py::test_query_large_page_small_rows has 30,000 items returned in a single page.

In the response JSON, all these items are returned in a single array "Items". Before this patch, we build the full response as a RapidJSON object before sending it. The problem is that unfortunately, RapidJSON stores arrays as contiguous allocations. This results in large contiguous allocations in workloads that scan many small items, and large contiguous allocations can also cause stalls and high tail latencies. For example, before this patch, running

    test/alternator/run --runveryslow \
        test_query.py::test_query_large_page_small_rows

reports in the log:

    oversized allocation: 573440 bytes.

After this patch, this warning no longer appears.
The patch solves the problem by collecting the scanned items not in a RapidJSON array, but rather in a chunked_vector<rjson::value>, i.e, a chunked (non-contiguous) array of items (each a JSON value). After collecting this array separately from the response object, we need to print its content without actually inserting it into the object - we add a new function print_with_extra_array() to do that.

The new separate-chunked-vector technique is used when a large number (currently, >256) of items were scanned. When there is a smaller number of items in a page (this is typical when each item is longer), we just insert those items in the object and print it as before.

Beyond the original slow test that demonstrated the oversized allocation (which is now gone), this patch also includes a new test which exercises the new code with a scan of 700 (>256) items in a page - but this new test is fast enough to be permanently in our test suite and not a manual "veryslow" test as the other test.

Fixes #23535

The stalls caused by large allocations was seen by actual users, so it makes sense to backport this patch. On the other hand, the patch while not big is fairly intrusive (modifies the nomal Scan and Query path and also the later patches do some cleanup of additional code) so there is some small risk involved in the backport.

nyh commented:
@avikivity I marked this patch "Fixes https://github.com/scylladb/scylladb/issues/23535" but it's not entirely clear to me what #23535 is about so I don't know if the "Fixes" is fair:

The original claim you made in that issue, that Alternator somehow misuses paging or forgets to limit it, was not correct as far as I understand. What we learned in the discussion in that issue was that the default 1MB page size Alternator (correctly) uses can result in a large number of tiny rows (30,000 in the test we already had) which caused a large contiguous allocation of a RapidJSON array. It is this contiguous allocation - and **only** this contiguous allocation - which this patch fixes. After this patch, a query returning 30,000 tiny items in one page, has no "oversized allocation" warnings.

Do you think that's enough for marking #23535 fixed, or is there anything else you think needs to be fixed before that issue is closed.

nyh commented on alternator/executor.cc:
```diff
@@ -156,6 +156,50 @@ json::json_return_type make_streamed(rjson::value&& value) {
+// rjson::value because RapidJSON does contiguous allocations for arrays
+// which we want to avoid for potentially long arrays in Query/Scan responses
+// (see #23535).
+// If we can ever fix RapidJSON (or a replacement) to avoid contiguous
```
I mentioned the possibility of replacing RapidJSON (see "or a replacement"), I'll reconsider the wording and add a reference to that issue.

nyh replied:
By the way, soon all the references to issues in our source code will become outdated garbage. Very sad. It was apparently decided to migrate to JIRA and to NOT to keep the old issue numbers when doing the migration.

nyh replied:
There is no reason to be optimistic. A couple of years from now, if we're lucky, nobody will remember what "#123" refers to.  If we're less lucky, we'll close down github issues and the old issues will not be available at all.

If we wanted to keep these numbers working, we have a way. But the official decision was not to keep the old numbers :-(

nyh commented on alternator/executor.cc:
```diff
@@ -156,6 +156,50 @@ json::json_return_type make_streamed(rjson::value&& value) {
+    // Need to make a copyable version of value. Gah.
+    auto rs = make_shared<rjson::value>(std::move(value));
+    auto ns = make_shared<std::string>(std::move(array_name));
+    auto as = make_shared<utils::chunked_vector<rjson::value>>(std::move(array));
```
Again, all these comments apply to the code I copied from :-(
I'll spend some time now to completely rewrite this function, and then think if I I should also change the original function I copied...

nyh replied:
Done. In the new version there is a second patch revamping this pre-existing code (which after this patch has two copies) to no longer need to use all these shared pointers, because it uses noncopyable function/

nyh commented:
> @nyh - ping - let's try to get this in soon.

I'm working on this now. Both reviewers did not like a particular piece of code that I actually copied from pre-existing code and was admittedly ugly because it was built on an inappropriate Seastar feature (Seastar's "json_return_type")

nyh commented on alternator/executor.cc:
```diff
@@ -156,6 +156,50 @@ json::json_return_type make_streamed(rjson::value&& value) {
+// rjson::print_with_extra_array() which it calls.
+json::json_return_type make_streamed_with_extra_array(rjson::value&& value,
+    std::string array_name, utils::chunked_vector<rjson::value>&& array) {
+    // CMH. json::json_return_type uses std::function, not noncopyable_function.
```
Done. Will appear in the next version of this PR as a second patch.

nyh commented on alternator/executor.cc:
```diff
@@ -156,6 +156,50 @@ json::json_return_type make_streamed(rjson::value&& value) {
+            // just log + rethrow. It is probably better the HTTP server closes connection
+            // abruptly or something...
+            ex = std::current_exception();
+            elogger.error("Exception during streaming HTTP response: {}", ex);
```
I'll add a third patch to remove these log messages. There is no real need to log these exceptions. The most likely cause of these exceptions will be a broken connection.

nyh commented on alternator/executor.cc:
```diff
@@ -4859,7 +4919,7 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
     auto paging_state = rs->get_metadata().paging_state();
     bool has_filter = filter;
-    auto [items, size] = co_await describe_items(*selection, std::move(rs), std::move(attrs_to_get), std::move(filter));
+    auto [items, actual_items, size] = co_await describe_items(*selection, std::move(rs), std::move(attrs_to_get), std::move(filter));
```
I don't know. "items" is the object which contains all the items' descriptions (count, etc.), and "actual items" contains the actual items.
I'll rename them to items_descr and items.

nyh commented on alternator/executor.cc:
```diff
@@ -4868,6 +4928,22 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
+        // let's just insert them into the JSON object and print the
+        // full JSON normally.
+        rjson::value items_json = rjson::empty_array();
+        for (auto& item : *actual_items) {
```
In some sense, yes. HOWEVER, the loop only runs when there is a relatively small number of items (currently this 256). Hopefully when there is a small number of items, the cost of this loop is negligible.

One of the benefits of doing this is that it allows me to, after building the output, decide if it's "big" or not and potentially output everything together instead of streaming. I was hoping it will make everything faster when the responses are small.

Maybe the number 256 was excessive and I should use a smaller number.

nyh commented on test/alternator/test_query.py:
```diff
@@ -637,3 +637,23 @@ def test_query_large_page_small_rows(test_table_sn):
+    i = 0
+    for item in got_items:
+        assert item == {'p': p, 'c': i}
+        i += 1
```
Ok, done.

nyh replied:
Note it's N == i, not i+1.

nyh commented on utils/rjson.cc:
```diff
@@ -280,6 +280,36 @@ future<> print(const rjson::value& value, seastar::output_stream<char>& os, size
+        for (const rjson::value& v : array) {
+            v.Accept(writer);
+        }
+        writer.EndArray();
```
You're right. Done

nyh commented on utils/rjson.hh:
```diff
@@ -158,6 +159,21 @@ std::string print(const rjson::value& value, size_t max_nested_level = default_m
+    std::string_view array_name,
+    const utils::chunked_vector<rjson::value>& array,
+    seastar::output_stream<char>& os,
+    size_t max_nested_level  = default_max_nested_level);
```
Fixed.

nyh commented:
Pushed a new version addressing many of the reviewers' comments (see my responses on each of these comments).

The most important change is the second patch, which rewrites the ugly and hard to understand make_streamed() and make_streamed_with_extra_array(). I dropped Alternator's use of the ill-suited seastar::json_return_type and replaced it by something more suited for Alternator's needs, and a lot of the code, especially make_streamed() became much shorter without those strange copies and extra shared pointers needed before.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/)<br> :small_blue_diamond: alternator | :white_check_mark: | spider2.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/)<br> :small_blue_diamond: alternator | :white_check_mark: | spider7.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18005/)<br> :small_blue_diamond: alternator | :white_check_mark: | spider3.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5782/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17060/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8976/) | :white_check_mark: |  |

#### Build Details:
- Duration: 1 hr 51 min

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -1212,7 +1214,7 @@ future<executor::request_return_type> executor::list_tags_of_resource(client_sta
     }
 
-    return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
+    return make_ready_future<executor::request_return_type>(rjson::print(std::move(ret)));
```
`co_return rjson::print(std::move(ret))` should also work and is shorter. If you follow this suggestion, line 1198 will also require a change into `co_return`.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -2417,7 +2419,7 @@ static future<executor::request_return_type> rmw_operation_return(rjson::value&&
         rjson::add(ret, "Attributes", std::move(attributes));
     }
-    return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
+    return make_ready_future<executor::request_return_type>(rjson::print(std::move(ret)));
```
`co_return`?

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -4256,7 +4258,7 @@ future<executor::request_return_type> executor::get_item(client_state& client_st
         _stats.api_operations.get_item_latency.mark(std::chrono::steady_clock::now() - start_time);
         uint64_t rcu_half_units = 0;
-        auto res = make_ready_future<executor::request_return_type>(make_jsonable(describe_item(schema, partition_slice, *selection, *qr.query_result, std::move(attrs_to_get), add_capacity, rcu_half_units)));
+        auto res = make_ready_future<executor::request_return_type>(rjson::print(describe_item(schema, partition_slice, *selection, *qr.query_result, std::move(attrs_to_get), add_capacity, rcu_half_units)));
```
~make_ready_future<executor::request_return_type>~ and then `co_return` at the end?

nyh replied:
I don't remember why, but this function is co-routinized by uses a then() in one place. Which I agree is kind of ugly but I don't think I should change it in this patch. I'll do it in a fourth patch I'll add to this series (although arguably, it's not even related to this series).

ScyllaPiotr reviewed (approved):
Nice one!

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -4868,6 +4928,22 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
         cql_stats.filtered_rows_matched_total += size;
     }
+    if (actual_items) {
+        if (actual_items->size() >= 256) {
```
I just noticed that I have re-raised this in my latest review.
You suggest to do this later, but I'm afraid we will forget this. It does not need to be configurable, but at least let it be a named constant. About the name, what about `max_resp_item_count_printed_by_rapidjson_natively`?

nyh commented on alternator/executor.cc:
```diff
@@ -4730,13 +4777,25 @@ class describe_items_visitor {
+// and "ScannedCount", but *not* "Items" - that is returned separately
+// as a chunked_vector to avoid large contiguous allocations which
+// RapidJSON does of its array. The caller should add "Items" to the
+// returned object if needed, or print it separately.
```
ok, I'll add this word.

nyh commented on alternator/executor.cc:
```diff
@@ -4882,19 +4900,35 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
+            // There are many items, better print the JSON and the array of
+            // items (opt_items) separately to avoid RapidJSON's contiguous
+            // allocation of arrays.
+            co_return executor::request_return_type(make_streamed_with_extra_array(std::move(items_descr), "Items", std::move(*opt_items)));
```
Good catch. I'll remove this unnecessary cast.
I think this is a leftover from the days of continuations (not coroutines) where if a continuation has multiple returns (not co_returns) they need to have explicitly the same type and if it wasn't exactly the same type (e.g., string vs function) I had to explicitly cast in all return points. With co_return, this is no longer the case, which is nice.

I think I'll do this cleanup in the second patch, not this one, though - because the second patch cleans up that make_jsonable thing in the pre-existing code, and I want to remove all those unnecessary request_return_type things in one patch.

nyh commented on alternator/executor.cc:
```diff
@@ -5531,7 +5565,7 @@ future<executor::request_return_type> executor::list_tables(client_state& client
     }
 
-    return make_ready_future<executor::request_return_type>(make_jsonable(std::move(response)));
+    return make_ready_future<executor::request_return_type>(rjson::print(std::move(response)));
```
I'll do this too in the new fourth patch

nyh commented on alternator/executor.hh:
```diff
@@ -169,7 +146,19 @@ class executor : public peering_sharded_service<executor> {
+    // The body_writer is used for streaming responses, where the response body
+    // is written in chunks to the output_stream. This allows for efficient
+    // handling of large responses without needing to allocate a large buffer
+    // in memory.
```
No, I don't think so. We've been using streamed responses long before #23535 - e.g., a BatchGetItem of several large items and other places.

nyh commented:
Rebased and addressed a bunch of review comments from @ScyllaPiotr. Many of the comments lamented ugly casting in return statements that could be addressed by using co_return instead. Because changing this required changing additional irrelevant things in the same functions (you can't have a single function doing co_return in one place, return in another place) I did this cleanup as a fourth patch.

@ScyllaPiotr let's get this series reviewed and merged before it grows further (it started as one patch, now it's 4... :-)).

nyh commented:
Restarted the dtest run (I wasn't sure if I should restart the whole build again or just the dtest?)

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -4815,6 +4832,12 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
+// returning a large number of items as a single rapidjson array, and use
+// a chunked_vector instead. The following constant is an arbitrary cutoff
+// point for when to switch from a rapidjson array to a chunked_vector.
+static constexpr int max_items_for_rapidjson_array = 256;
```
Nice and clear explanation :+1:

nyh commented:
> Restarted the dtest run (I wasn't sure if I should restart the whole build again or just the dtest?)

I don't see new test results. I'll rebase to get the CI running from scratch.

mykaul commented:
@scylladb/scylla-maint - please review for inclusion.

nyh commented:
Users are still encountering this problem on older releases. Above we only did the backport to 2025.3 and the backports to 2025.2 and 2025.1 which had conflicts were not done. I'll need to do this now.
